### ISSUE TITLE
Debug Service Continues to Fail (Refer to issue #2110)

### ISSUE THREAD

**Issue by Nick1497** [codefori/vscode-ibmi#2114]
Type: <b>Bug</b>

Connect to IBM i system > Start the debug server > Attempt to start the debug service in which the issue occurs at this step. It is affecting my entire team who use this tool, we all receive the same error message as well.

![image](https://github.com/codefori/vscode-ibmi/assets/172528062/16619b62-0f03-466b-ab2e-86817eabf592)

Opened new issue as I did not see any option to re-open the old one. The first suggestion in #2110 was to wipe the certs and generate new ones and that did not fix the problem. The second troubleshooting step had an unrelated error message to mine. Seeing if anyone has any other suggestions. 

Extension version: 2.10.1
VS Code version: Code 1.90.0 (89de5a8d4d6205e5b11647eb6a74844ca23d2573, 2024-06-04T19:33:54.889Z)
OS version: Windows_NT x64 10.0.19045
Modes:

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|AMD Ryzen 5 PRO 7530U with Radeon Graphics      (12 x 1996)|
|GPU Status|2d_canvas: enabled<br>canvas_oop_rasterization: enabled_on<br>direct_rendering_display_compositor: disabled_off_ok<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>opengl: enabled_on<br>rasterization: enabled<br>raw_draw: disabled_off_ok<br>skia_graphite: disabled_off<br>video_decode: enabled<br>video_encode: enabled<br>vulkan: disabled_off<br>webgl: enabled<br>webgl2: enabled<br>webgpu: enabled|
|Load (avg)|undefined|
|Memory (System)|15.31GB (5.00GB free)|
|Process Argv|--crash-reporter-id f363ceb0-20c6-49c7-80ad-3a25bbb3f05d|
|Screen Reader|no|
|VM|0%|
</details><details>
<summary>A/B Experiments</summary>

```
vsliv368cf:30146710
vspor879:30202332
vspor708:30202333
vspor363:30204092
vscoreces:30445986
vscod805cf:30301675
binariesv615:30325510
vsaa593cf:30376535
py29gd2263:31024239
c4g48928:30535728
azure-dev_surveyone:30548225
a9j8j154:30646983
962ge761:30959799
pythongtdpath:30769146
welcomedialog:30910333
pythonidxpt:30866567
pythonnoceb:30805159
asynctok:30898717
pythontestfixt:30902429
pythonregdiag2:30936856
pythonmypyd1:30879173
2e7ec940:31000449
pythontbext0:30879054
accentitlementst:30995554
dsvsc016:30899300
dsvsc017:30899301
dsvsc018:30899302
cppperfnew:31000557
dsvsc020:30976470
pythonait:31006305
jchc7451:31067544
chatpanelc:31048052
dsvsc021:30996838
bdiig495:31013172
pythoncenvpt:31062603
a69g1124:31058053
dvdeprecation:31068756
pythonprt:31056678
dwnewjupytercf:31046870
26j00206:31048877

```

</details>

**Comment by sebjulliand** [comment]
Two questions for you:
- What does this query returns when run on your system?
```sql
With JAVA_PTF_GROUP as (
  SELECT PTF_GROUP_TARGET_RELEASE OS,
  Case PTF_GROUP_TARGET_RELEASE
    When 'V7R4M0' Then 'SF99665'
    When 'V7R3M0' Then 'SF99725'
    When 'V7R2M0' Then 'SF99716'
    When 'V7R1M0' Then 'SF99572' End PTF_GROUP
FROM QSYS2.GROUP_PTF_INFO
WHERE PTF_GROUP_DESCRIPTION = 'TECHNOLOGY REFRESH'
AND PTF_GROUP_STATUS = 'INSTALLED'
LIMIT 1)
Select Max(OS) OS,
PTF_GROUP_NAME,
Max(PTF_GROUP_LEVEL) "CURRENT LEVEL"
From  QSYS2.GROUP_PTF_INFO
Inner Join JAVA_PTF_GROUP On PTF_GROUP = PTF_GROUP_NAME
Group by PTF_GROUP_NAME;
```

- What's the output of `java -version` when run in batch?

**Comment by Nick1497** [comment]
- Return of query: 
![image](https://github.com/codefori/vscode-ibmi/assets/172528062/71d4b38c-dd30-42c3-8132-005c09906b1f)

- The system is running 11.0.8 but JAVA_HOME that the debugger is referencing is java version 1.8.0_341

**Comment by Nick1497** [comment]
This links seems like it may be useful https://www.ibm.com/support/pages/support-hmacpbesha256-algorithm-app-connect-enterprise. Will investigate this to see if it will assist.

**Comment by sebjulliand** [comment]
I looked into this a bit and every time, it boils down to "Java doesn't support this algorithm". I would understand if the keystore was generated by Java, but we use `openssl` to generate it; and it worked fine on every LPAR I tried it on (i.e. Java 8 had no issue opening the keystore).

What's the output of these commands for you?
```
openssl version
```
```
which openssl
```

**Comment by mkwan01** [comment]
IBM i Debug v2.0.1 was just released. Debug v2 uses Java 11 instead of Java 8. This Java 8 specific error is unlikely to happen in Java 11.

**Comment by Nick1497** [comment]
Nice, when the system is back up I will give this a go! Updates to come.

**Comment by Nick1497** [comment]
I've updated to IBM i Debug v2.0.1 but, when I boot up the system and such the debugger still says 1.0.0: 
![image](https://github.com/codefori/vscode-ibmi/assets/172528062/17fa9f79-b38f-4b6e-bf30-0f2f8f8f4e60)

Any way to get it to use the new version?

**Comment by sebjulliand** [comment]
> I've updated to IBM i Debug v2.0.1 but, when I boot up the system and such the debugger still says 1.0.0:
> 
> ![image](https://private-user-images.githubusercontent.com/172528062/341401611-17fa9f79-b38f-4b6e-bf30-0f2f8f8f4e60.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTg4ODYyODIsIm5iZiI6MTcxODg4NTk4MiwicGF0aCI6Ii8xNzI1MjgwNjIvMzQxNDAxNjExLTE3ZmE5Zjc5LWIzOGYtNGI2ZS1iZjMwLTBmMmY4ZjhmNGU2MC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNjIwJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDYyMFQxMjE5NDJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT01YzgxNDAwYzE4MmNlOGRhMTQ4N2UyZWY5NjJiNGI1ZGJjODE3YWIzY2I5NTgzOTBiMjI2NmM2MDgxNDdlYTQwJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.AFb-zl5mpL5sgMr3ejvtEiDqX8EAVU45TTmgkTGdHII)
> 
> Any way to get it to use the new version?

You need to install the host update too to get Debugger v2.
![image](https://github.com/codefori/vscode-ibmi/assets/11096890/846b33ab-8202-4c3d-85f6-55e24edc9c72)

And now that I think about it, the debug service configuration will have to be updated to use Java 11, too.

**Comment by sebjulliand** [comment]
> Noted, I will look into getting the new PTF next week, and will also change the debugger config. Updates to come.
Nice! To update the configuration, open it from the Debugger view:
![image](https://github.com/codefori/vscode-ibmi/assets/11096890/0a476ade-006c-4e94-9365-75ca75aec835)

Then scroll down and update the JAVA_HOME variable
![image](https://github.com/codefori/vscode-ibmi/assets/11096890/c21de8f4-22a4-4e62-b4f3-17a3bf86afa0)
Change it to:
```
JAVA_HOME=/QOpenSys/QIBM/ProdData/JavaVM/jdk11/64bit
```
Save the configuration and you should be good to go.
Alternatively, you can delete the existing certificate and let Code for IBM i create a new one; it will update the configuration as well.

**Comment by Nick1497** [comment]
@sebjulliand to answer your question: 

> I looked into this a bit and every time, it boils down to "Java doesn't support this algorithm". I would understand if the keystore was generated by Java, but we use `openssl` to generate it; and it worked fine on every LPAR I tried it on (i.e. Java 8 had no issue opening the keystore).
> 
> What's the output of these commands for you?
> 
> ```
> openssl version
> ```
> 
> ```
> which openssl
> ```


<b>Openssl version</b>: OpenSSL 3.0.10 1 Aug 2023 (Library: OpenSSL 3.0.10 1 Aug 2023)

<b>which openssl</b>: /QOpenSys/usr/bin/openssl

Our system hardware was upgraded and there has been some issues, so for the time being I am not able to apply the new PTF to allow me to use the new debugger (2.0). Would be nice if I could get version 1.0 working while I wait.

**Comment by sebjulliand** [comment]
Maybe the output from `which openssl` is a good hint. Your openssl version seems OK but `which` points to `/QOpenSys/usr/bin` instead of `/QOpenSys/pkgs/bin`.

Can you check your PATH variable?
```
echo $PATH
```

I assume that `/QOpenSys/usr/bin` must appear before `/QOpenSys/pkgs/bin`. If it's the case, then try to put `/QOpenSys/pkgs/bin` in first position in the `.bashrc` file in your home directory.

Then re-open a shell session and check that `which openssl` answers `/QOpenSys/pkgs/bin/openssl` this time.
Then generate the certificate again.

If it still doesn't work, I'll have one more trick up my sleeve (or maybe two...).

**Comment by Nick1497** [comment]
<b>echo $PATH</b> ouput: /usr/bin:.:/QOpenSys/usr/bin

And when I check my <b>.bashrc</b> in the following directory home/myProfileName it has the current contents: 
# Generated by Code for IBM i
export PATH=/QOpenSys/pkgs/bin:$PATH

So it seems like it is already setup how it should be?

**Comment by sebjulliand** [comment]
Try this now:
- Download this zip
- Unzip it
- Install the vsix found inside. It's a special build that includes the `-legacy` flag when `openssl` is invoked to generate the PKCS12 file.
- Connect to your LPAR
- Delete any existing certificate
- Generate a new one
- Start the debug service

I tried it on my system and it worked (with or without the `-legacy` flag).
Let me know how it goes for you after you generated a certificate with this `2.11.4-dev.0` build.

**Comment by Nick1497** [comment]
I installed the vsix file and removed the certificates. When trying to re-generate them I get this error (where the blacked out portion is the DNS for our system): 

![image](https://github.com/codefori/vscode-ibmi/assets/172528062/f97991ee-7db2-4436-9183-f906ee551a26)

**Comment by sebjulliand** [comment]
After having a close look, it turns out your `PATH` makes you use `openssl` provided by yum instead of `/QOpenSys/usr/bin/openssl`. This causes some issues, obviously 😅 

Here is another vsix generated from this PR: https://github.com/codefori/vscode-ibmi/pull/2151
[code-for-ibmi-2.11.5-dev.0.vsix.zip](https://github.com/user-attachments/files/16101619/code-for-ibmi-2.11.5-dev.0.vsix.zip)


It will force the use of `/QOpenSys/usr/bin/openssl` when generating the certificates. This should take care of your issue once and for all.

Let me know how it goes!

**Comment by Nick1497** [comment]
Installed it and while the certificates could be generated this time. The original error still persists (I do enjoy how you added the output of the error message to VS Code!): 

![image](https://github.com/codefori/vscode-ibmi/assets/172528062/b994c538-c4f6-483b-ad9d-0884c988f22e)

**Comment by sebjulliand** [comment]
I'll need to check two thing related to your environment:
- The content of your `/QOpenSys/QIBM/ProdData/JavaVM/jdk80/64bit/jre/lib/security/java.security` file
- The output of the `env` command when run in a PASE terminal

Also, while we're at it, from a PASE terminal, run this:
```
/QOpenSys/usr/bin/openssl genrsa -out test_cert.key 2048 && /QOpenSys/usr/bin/openssl req -new -key test_cert.key -out test_cert.csr -subj '/CN=localhost' && /QOpenSys/usr/bin/openssl x509 -req -in test_cert.csr -signkey test_cert.key -out test_cert.crt -days 1095 -sha256 -req && /QOpenSys/usr/bin/openssl pkcs12 -export -out test_keystore.pfx -inkey test_cert.key -in test_cert.crt  -password pass:password && rm test_cert.*
```
This will create a test_keystore.pfx file in the current directory on the IFS. Please retrieve it, put it in a zip file and attach it in a reply.

Thanks!

**Comment by Nick1497** [comment]
[Java Security Contents.pdf](https://github.com/user-attachments/files/16110544/Java.Security.Contents.pdf)

<b>Env Command:</b> (blacked out company/personal info)
![image](https://github.com/codefori/vscode-ibmi/assets/172528062/5547443f-4877-4f2e-a944-b774322fb755)
![image](https://github.com/codefori/vscode-ibmi/assets/172528062/22b5cb16-92fc-4b44-afa9-2197c7ab32f8)
[test_keystore.zip](https://github.com/user-attachments/files/16110689/test_keystore.zip)

**Comment by Nick1497** [comment]
Any updates on this @sebjulliand?

**Comment by sebjulliand** [comment]
No...I compared your file with the ones I have on two LPAR running the service and couldn't spot any difference.
Nothing useful in the env output either.

The only thing I can say is that on the 7.3 box that runs the service, current level is 28:
![image](https://github.com/codefori/vscode-ibmi/assets/172528062/71d4b38c-dd30-42c3-8132-005c09906b1f)

If you can, ask to update SF99725 to the latest level. It shouldn't hurt and it may solve this issue.

**Comment by sebjulliand** [comment]
Also, if you can, attach these files if they exist.

Debug service log:
/QIBM/UserData/IBMIDEBUGSERVICE/DebugService_log.txt

Debug service Eclipse instance log:
/QIBM/UserData/IBMIDEBUGSERVICE/startDebugService_workspace/.metadata/.log

Thanks!

**Comment by Nick1497** [comment]
I will look into upgrading the version you mentioned. I did not see anything for the <b>Debug Service Log</b> but here is the other .log file: [.log](https://github.com/user-attachments/files/16228603/default.log)

**Comment by Nick1497** [comment]
Yeah our team can't really apply any PTFs right now, so upgrading the version won't be possible currently. When we are able to apply PTFs I will get the ones that allow me to use V2 of the debugger.
